Crashing on obfuscated script

### How can we reproduce the crash?

I'm trying to write a JavaScript obfuscation tool and one of the outputs in my tests is crashing bun.

https://github.com/imkunet/bun_crash_obfuscated_code

### Relevant log output

```shell
~/code/kunet/js/debaucherator main +1 !16 ?6 🌸 bun /Users/kunet/code/kunet/js/debaucherator/tests/4_template.obf.js                  2.6s ▼  impure
============================================================
Bun v1.2.5 (013fdddc) macOS Silicon
macOS v15.3.2
Args: "bun" "/Users/kunet/code/kunet/js/debaucherator/tests/4_template.obf.js"
Features: jsc tsconfig tsconfig_paths
Builtins: "bun:main"
Elapsed: 7ms | User: 14ms | Sys: 9ms
RSS: 31.39MB | Peak: 31.39MB | Commit: 1.07GB | Faults: 38

panic(main thread): Segmentation fault at address 0x58
Crashed while visiting /Users/kunet/code/kunet/js/debaucherator/tests/4_template.obf.js
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.2.5/Ma1013fdddMgggQ_2hvzS2hvzSut5qR287qRm1srRmukrR287qRmx+qRm1srRA2AwF

zsh: trace trap  bun /Users/kunet/code/kunet/js/debaucherator/tests/4_template.obf.js
```

### Stack Trace (bun.report)

Bun v1.2.5 ([`013fddd`](<https://github.com/oven-sh/bun/tree/013fdddc6ed18bc849614ccb37a296fa0f69a5db>)) on macos aarch64 [AutoCommand]

Segmentation fault at address 0x00000058

- *1 unknown/js code*
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`
- `js_parser.NewParser_`

Features: tsconfig, tsconfig\_paths, jsc

<!-- from bun.report: 4Mjr9nveiidYDZpUohcV -->

<sub>Sentry Issue: <strong><a href="https://bun-p9.sentry.io/issues/6294958575/">BUN-CKT</a></strong></sub>

Looks like a minification bug

```zig
Bun Debug v1.2.6 (6e7c1d7e) macOS Silicon
macOS v15.3.1
CPU: neon fp aes crc32 atomics
Args: "bun-debug" "build" "--minify-syntax" "a.js"
Features: bunfig tsconfig(2)
Elapsed: 71ms | User: 81ms | Sys: 49ms
RSS: 0.26GB | Peak: 0.26GB | Commit: 15.99MB | Faults: 719

panic: Scope mismatch while visiting
428 |               return "".concat(絃)[function () {
                                        ^
debug: Expected this scope (.function_args)
    at /Users/jarred/Code/bun/a.js:428:35

783 |     } while ('\u9b63\u9b94\u54cb\u6618' === function () {
                                                  ^
debug: Found this scope (.function_args)
    at /Users/jarred/Code/bun/a.js:783:45

Crashed while visiting /Users/jarred/Code/bun/a.js
/Users/jarred/Code/bun/src/output.zig:484:24: 0x10e3e9f3f in panic__anon_1077808 (bun-debug)
        std.debug.panic(comptime prettyFmt(fmt, true), args);
                       ^
/Users/jarred/Code/bun/src/js_parser.zig:14671:25: 0x10783588b in panicLoc__anon_972124 (bun-debug)
            Output.panic(fmt ++ "\n{s}", args ++ .{panic_buffer[0..panic_stream.pos]});
                        ^
/Users/jarred/Code/bun/src/js_parser.zig:14648:23: 0x10e3981c3 in panic__anon_883897 (bun-debug)
            p.panicLoc(fmt, args, null);
                      ^
/Users/jarred/Code/bun/src/js_parser.zig:6979:24: 0x107258faf in pushScopeForVisitPass (bun-debug)
                p.panic("Scope mismatch while visiting", .{});
                       ^
/Users/jarred/Code/bun/src/js_parser.zig:16049:36: 0x107c86bf7 in visitFunc (bun-debug)
            p.pushScopeForVisitPass(.function_args, open_parens_loc) catch unreachable;
                                   ^
/Users/jarred/Code/bun/src/js_parser.zig:17607:42: 0x107a6e19b in visitExprInOut (bun-debug)
                    const old_inside_async_arrow_fn = p.fn_only_data_visit.is_inside_async_arrow_fn;
                                         ^
/Users/jarred/Code/bun/src/js_parser.zig:17233:49: 0x107a6b737 in visitExprInOut (bun-debug)
                    // even when minification is disabled, so that if we have an
                                                ^
/Users/jarred/Code/bun/src/js_parser.zig:16023:36: 0x107a66e87 in visitExprInOut (bun-debug)
            return p.visitExprInOut(expr, .{});
                                   ^
/Users/jarred/Code/bun/src/js_parser.zig:17233:49: 0x107a6b737 in visitExprInOut (bun-debug)
                    // even when minification is disabled, so that if we have an
                                                ^
/Users/jarred/Code/bun/src/js_parser.zig:16960:49: 0x107a68f9f in visitExprInOut (bun-debug)
                        if (define_data.call_can_be_unwrapped_if_unused and !p.options.ignore_dce_annotations) {
                                                ^

fish: Job 1, 'bun-debug build --minify-syntax…' terminated by signal SIGTRAP (Trace or breakpoint trap)
```

--- a/src/ast/visitExpr.zig
+++ b/src/ast/visitExpr.zig
@@ -579,58 +579,7 @@ pub fn VisitExpr(
                         e_.index = .{ .data = .{ .e_private_identifier = private }, .loc = e_.index.loc };
                     },
                     else => {
-                        const index = p.visitExpr(e_.index);
-                        e_.index = index;
-
-                        const unwrapped = e_.index.unwrapInlined();
-                        if (unwrapped.data == .e_string and
-                            unwrapped.data.e_string.isUTF8())
-                        {
-                            // "a['b' + '']" => "a.b"
-                            // "enum A { B = 'b' }; a[A.B]" => "a.b"
-                            if (p.options.features.minify_syntax and
-                                unwrapped.data.e_string.isIdentifier(p.allocator))
-                            {
-                                const dot = p.newExpr(
-                                    E.Dot{
-                                        .name = unwrapped.data.e_string.slice(p.allocator),
-                                        .name_loc = unwrapped.loc,
-                                        .target = e_.target,
-                                        .optional_chain = e_.optional_chain,
-                                    },
-                                    expr.loc,
-                                );
-
-                                if (is_call_target) {
-                                    p.call_target = dot.data;
-                                }
-
-                                if (is_delete_target) {
-                                    p.delete_target = dot.data;
-                                }
-
-                                return p.visitExprInOut(dot, in);
-                            }
-
-                            // Handle property rewrites to ensure things
-                            // like .e_import_identifier tracking works
-                            // Reminder that this can only be done after
-                            // `target` is visited.
-                            if (p.maybeRewritePropertyAccess(
-                                expr.loc,
-                                e_.target,
-                                unwrapped.data.e_string.data,
-                                unwrapped.loc,
-                                .{
-                                    .is_call_target = is_call_target,
-                                    // .is_template_tag = is_template_tag,
-                                    .is_delete_target = is_delete_target,
-                                    .assign_target = in.assign_target,
-                                },
-                            )) |rewrite| {
-                                return rewrite;
-                            }
-                        }
+                        e_.index = p.visitExprInOut(e_.index, .{});
                     },
                 }
 

--- a/src/ast/visitExpr.zig
+++ b/src/ast/visitExpr.zig
@@ -579,7 +579,59 @@ pub fn VisitExpr(
                         e_.index = .{ .data = .{ .e_private_identifier = private }, .loc = e_.index.loc };
                     },
                     else => {
-                        e_.index = p.visitExprInOut(e_.index, .{});
+                        const index = p.visitExpr(e_.index);
+                        e_.index = index;
+
+                        const unwrapped = e_.index.unwrapInlined();
+                        if (unwrapped.data == .e_string and
+                            unwrapped.data.e_string.isUTF8())
+                        {
+                            // "a['b' + '']" => "a.b"
+                            // "enum A { B = 'b' }; a[A.B]" => "a.b"
+                            if (p.options.features.minify_syntax and
+                                unwrapped.data.e_string.isIdentifier(p.allocator))
+                            {
+                                const dot = p.newExpr(
+                                    E.Dot{
+                                        .name = unwrapped.data.e_string.slice(p.allocator),
+                                        .name_loc = unwrapped.loc,
+                                        .target = e_.target,
+                                        .optional_chain = e_.optional_chain,
+                                    },
+                                    expr.loc,
+                                );
+
+                                if (is_call_target) {
+                                    p.call_target = dot.data;
+                                }
+
+                                if (is_delete_target) {
+                                    p.delete_target = dot.data;
+                                }
+
+                                // don't call visitExprInOut on `dot` because we've already visited `target` above!
+                                return dot;
+                            }
+
+                            // Handle property rewrites to ensure things
+                            // like .e_import_identifier tracking works
+                            // Reminder that this can only be done after
+                            // `target` is visited.
+                            if (p.maybeRewritePropertyAccess(
+                                expr.loc,
+                                e_.target,
+                                unwrapped.data.e_string.data,
+                                unwrapped.loc,
+                                .{
+                                    .is_call_target = is_call_target,
+                                    // .is_template_tag = is_template_tag,
+                                    .is_delete_target = is_delete_target,
+                                    .assign_target = in.assign_target,
+                                },
+                            )) |rewrite| {
+                                return rewrite;
+                            }
+                        }
                     },
                 }
 

--- a/test/bundler/transpiler/transpiler.test.js
+++ b/test/bundler/transpiler/transpiler.test.js
@@ -3551,6 +3551,19 @@ it("does not crash with 9 comments and typescript type skipping", () => {
   expect(exitCode).toBe(0);
 });
 
+it("does not crash with --minify-syntax and revisiting dot expressions", () => {
+  const { stdout, stderr, exitCode } = Bun.spawnSync({
+    cmd: [bunExe(), "-p", "[(()=>{})()][''+'c']"],
+    stdout: "pipe",
+    stderr: "pipe",
+    env: bunEnv,
+  });
+
+  expect(stderr.toString()).toBe("");
+  expect(stdout.toString()).toBe("undefined\n");
+  expect(exitCode).toBe(0);
+});
+
 it("runtime transpiler stack overflows", async () => {
   expect(async () => await import("./fixtures/lots-of-for-loop.js")).toThrow(`Maximum call stack size exceeded`);
 });

